### ISSUE TITLE
Improve error reporting when an operation fails

### ISSUE THREAD

**Issue by susanodd** [Signbank/Global-signbank#1331]
Instead of just try-except pass around error messages to report back. Also for transactions that fail. Also for file system errors.
This is non-trivial.

Kindly report use cases (including API) where something is going wrong but this isn't reported sufficiently.

**Comment by susanodd** [comment]
@rem0g are you busy trying to upload a zip file?

In the error log it was giving errors about the api token.

**Comment by rem0g** [comment]
Yes, sometimes video_upload returns whole django HTML template so my script doesnt know how to interprent that. It will keep for looping that same video file, so error reporting should be always in JSON format.

Sometimes i'm getting Failed to upload message, though there is no reason attached as why it happens. 

I will continue to run the script as some videos are uploaded succesfully, but I am not satisfied with how Signbank handles the video upload process so we will need another method to handle the videos as in the future we will have to process thousands of videos every week (happening now). 

I am getting fails with gloss with special chars for example: 
"DISNEY+-A"
"#WDD"

Also Signbank notifies the video is not h264 but yet the video is there on Signbank. 

@Woseseltops i think the video upload process needs a do-over; 
A seperate python file to retrieve video file via POST and process as is, it's not needed to check whether it's h264. If it doesnt play, it wont play and we will know that  one way or another.
JSON file so i can check every moment if the upload and placing the video has been successful.

**Comment by Jetske** [comment]
@rem0g what is a good way to report errors? Returning a JsonResponse with the error? I'm trying to add checks into the APIs for example sentences.

**Comment by Woseseltops** [comment]
Indeed @Jetske , something like this:

```python
return JsonResponse({'error': 'Invalid request'}, status=400)
```

In addition to that, using the correct error code would be even better:
* 400 Bad Request: The request was malformed.
* 401 Unauthorized: Authentication is needed.
* 402 Payment Required: Reserved for future use.
* 403 Forbidden: The server understood the request but refuses to authorize it.
* 404 Not Found: The requested resource could not be found.
* 405 Method Not Allowed: The request method is not supported for the resource.

@rem0g , I'll create a separate issue for your video upload suggestions!

**Comment by susanodd** [comment]
The error reporting I added for gloss updates reports type errors, for example choices that don't exist.

**Comment by susanodd** [comment]
@rem0g have you been able to use the NME API functions all right?

**Comment by susanodd** [comment]
> Indeed @Jetske , something like this:
> 
> ```python
> return JsonResponse({'error': 'Invalid request'}, status=400)
> ```
> 
> In addition to that, using the correct error code would be even better:
> 
> * 400 Bad Request: The request was malformed.
> * 401 Unauthorized: Authentication is needed.
> * 402 Payment Required: Reserved for future use.
> * 403 Forbidden: The server understood the request but refuses to authorize it.
> * 404 Not Found: The requested resource could not be found.
> * 405 Method Not Allowed: The request method is not supported for the resource.
> 
> @rem0g , I'll create a separate issue for your video upload suggestions!

Is there anything for "type checking"? Or do we make a new code for this?

**Comment by susanodd** [comment]
@rem0g do you know what specific API you called that reported that it failed but it did not?

**Comment by susanodd** [comment]
@rem0g I can see that there are Disney video files:

```
(env) root@signbank-new:/var/www/repo# ls -l ../writable/glossvideo/NGT/DI/DISNEY*
-rw-r--r-- 1 ubuntu www-data 1319804 Mar 25  2024 ../writable/glossvideo/NGT/DI/DISNEY-A-47512.bak.bak
-rw-r--r-- 1 ubuntu www-data 1899244 Sep 23 10:09 ../writable/glossvideo/NGT/DI/DISNEY-A-47512.mp4
-rw-r--r-- 1 ubuntu www-data 1805995 Sep 23 10:09 ../writable/glossvideo/NGT/DI/DISNEY-B-47513.mp4
-rw-r--r-- 1 ubuntu www-data 2260995 Sep 19 07:02 ../writable/glossvideo/NGT/DI/DISNEY-C-48427.mp4
-rw-r--r-- 1 ubuntu www-data 2442557 Sep 19 07:02 ../writable/glossvideo/NGT/DI/DISNEY-D-48428.mp4
(env) root@signbank-new:/var/www/repo# 
```

Of the above, DISNEY-C and DISNEY-D do not show on signbank, in spite of the video files existing.

**Comment by susanodd** [comment]
In the video history I can see you uploaded the DISNEY videos numerous times. (25 times or so)

**Comment by susanodd** [comment]
The above error is high priority!!!

**Comment by susanodd** [comment]
Up to here, the repairs commented about above:

- to video import (fixed path problem with gloss video object, do not check h264, do not return video link if import went wrong)
- report the dataset a gloss is in if not selected

have been deployed.

**Comment by Woseseltops** [comment]
> Is there anything for "type checking"? Or do we make a new status code for this? I guess it could fall under "Bad Request" but that is totally not helpful at all.

I would say:

```python
return JsonResponse({'error': 'Explanation of what went wrong type checking.'}, status=400)
```

**Comment by susanodd** [comment]
For the video upload when the video was not h264, those are imported now. Then you see for the gloss that the video cannot be displayed.
Do we need a command to find / search on these videos?
Or just assume if the user sees them in the gloss results that they'll notice?

https://signbank.cls.ru.nl/dictionary/gloss/48471

https://signbank.cls.ru.nl/dictionary/gloss/48424/

**Comment by susanodd** [comment]
> @rem0g 
> I am getting fails with gloss with special chars for example:
> "DISNEY+-A"
> "#WDD"

There is no match for WDD so I don't think it's the special character.
There is a #WWW (but no video)

https://signbank.cls.ru.nl/dictionary/gloss/49509/

**Comment by vanlummelhuizen** [comment]
Some things I found on the live server:

There are no GlossVideo objects for Gloss 48427 or for the video file:

```python
>>> GlossVideo.objects.filter(gloss_id=48427)
<QuerySet []>
>>> GlossVideo.objects.filter(videofile__contains='48427')
<QuerySet []>
```

In the logs (`/var/www/writable/logs/signbank.uwsgi.log`) I found the following and similar errors:

```log
Internal Server Error: /dictionary/upload_zipped_videos_folder_json/5/
Traceback (most recent call last):
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/repo/signbank/api_token.py", line 69, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/repo/signbank/api_interface.py", line 323, in upload_zipped_videos_folder_json
    with urllib.request.urlopen(zipped_file_url) as response:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 215, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 515, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 532, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 492, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 1392, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 1344, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.12/http/client.py", line 1331, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.12/http/client.py", line 1342, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python3.12/http/client.py", line 1175, in putrequest
    self._validate_path(url)
  File "/usr/lib/python3.12/http/client.py", line 1276, in _validate_path
    raise InvalidURL(f"URL can't contain control characters. {url!r} "
http.client.InvalidURL: URL can't contain control characters. '/DISNEY -B.zip' (found at least ' ')
[pid: 813|app: 0|req: 27705/135881] 10.208.155.251 () {60 vars in 1232 bytes} [Thu Sep 19 00:00:15 2024] GET /dictionary/upload_zipped_videos_folder_json/5/?file=https://www.gomerotterspeer.nl/DISNEY+-B.zip => generated 139367 bytes in 109 msecs (HTTP/1.1 500) 5 headers in 178 bytes (1 switches on core 1)
```

In this case it is the plus sign in the name of the zip file that led to this error.

**Comment by susanodd** [comment]
Hmmm.

`http.client.InvalidURL: URL can't contain control characters. '/DISNEY -B.zip' (found at least ' ')`

Now you're just trying to break it. :) The zip file should be an archive for NGT, not just one gloss.
It gets the gloss from the video files nested several levels in, NGT/nld/DISNEY-B.mp4

**Comment by susanodd** [comment]
I removed the videos that were in writable instead of writable/NGT/XX
I removed the gloss video objects that linked to them.
(There were thousands of videos and png files in the "writable" folder.)

**Comment by susanodd** [comment]
> There are no GlossVideo objects for Gloss 48427 or for the video file:
> 
> >>> GlossVideo.objects.filter(gloss_id=48427)
> <QuerySet []>
> >>> GlossVideo.objects.filter(videofile__contains='48427')
> <QuerySet []>

What is this meant to point out?
Yes, the gloss has no video.

**Comment by susanodd** [comment]
Here is what the video looks like if it's not h264 after the probe was removed. (That was a video that was stuck in the import folder because it had the wrong format.)

![video-niet-h264-mp4](https://github.com/user-attachments/assets/238d7668-52a5-4567-8030-5f7e8c694004)

**Comment by susanodd** [comment]
Fixed the Invalid URL problem. It is already made invalid before the code gets it.

**Comment by susanodd** [comment]
These files are still there:

```
(env) root@signbank-new:/var/www/repo# ls -l ../writable/glossvideo/NGT/DI/DISNEY*
-rw-r--r-- 1 ubuntu www-data 1319804 Mar 25  2024 ../writable/glossvideo/NGT/DI/DISNEY-A-47512.bak.bak
-rw-r--r-- 1 ubuntu www-data 1899244 Sep 23 10:09 ../writable/glossvideo/NGT/DI/DISNEY-A-47512.mp4
-rw-r--r-- 1 ubuntu www-data 1805995 Sep 23 10:09 ../writable/glossvideo/NGT/DI/DISNEY-B-47513.mp4
-rw-r--r-- 1 ubuntu www-data 2260995 Sep 19 07:02 ../writable/glossvideo/NGT/DI/DISNEY-C-48427.mp4
-rw-r--r-- 1 ubuntu www-data 2442557 Sep 19 07:02 ../writable/glossvideo/NGT/DI/DISNEY-D-48428.mp4
(env) root@signbank-new:/var/www/repo# 
```

But there are no objects pointing to these files.
What to do? Make a command to locate files without GlossVideo objects pointing to them?
@Jetske made such a command several years ago to locate such files because this has already been a problem in the past.
Some older code looked for a file if there was no object.
I guess it could be lazy, so if there is no GlossVideo object for the gloss, but there is a file, it could create an object for the file.
These files are in the correct location.

**Comment by susanodd** [comment]
Some video code in add_video (normal videos):

https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/dictionary/models.py#L2408-L2409

I couldn't get the code to use this 'else' branch.
This branch looks like it doesn't save the new GlossVideo object. But it may have already uploaded a video to "upload_to".

**Comment by vanlummelhuizen** [comment]
> Some video code in add_video (normal videos):
> 
> https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/dictionary/models.py#L2408-L2409

I found this code came into existence in this commit: https://github.com/Signbank/Global-signbank/commit/110435a67af792e4f77066c14ae9782a9c767393 Perhaps that may help you.

---

> this latter one is in code that is not accessible anymore (there is no URL that goes to this view function). It catches the result of the method. But this code is not called. It is the previous version of the Media Manager. All the urls have been removed. The function ought to be deleted. There are more of these hidden urls for other functionality. It was not clear whether some code is used by ASL.

I would say that this is dead code then and it can be removed.

---

> @vanlummelhuizen any ideas?

At this moment: no.

**Comment by susanodd** [comment]
So that commit is of code that is no longer accessible.

**Comment by Jetske** [comment]
@vanlummelhuizen I also tried api_create_gloss to test with Swagger and failed ;) it's already fixed in the pull request for api's for annotated/example sentences (if you have time feel free to review it and merge). 

I didn't try any zip files yet.

**Comment by susanodd** [comment]
Oh, the zip file was from before the API token, so that code was never modified.
Normally, if you're already logged in, it just works.

There is a "twin" command for actually "importing" the videos that are in the zip archive.
It works basically the same way as the Manage Media page.

@Woseseltops wrote the swagger code. I haven't tried it.

If the zip file has been uploaded (by the command you show above) then it's in the same place as the other zip files you can see on Manage Media.

**Comment by susanodd** [comment]
> @vanlummelhuizen I also tried api_create_gloss to test with Swagger and failed ;) it's already fixed in the pull request for api's for annotated/example sentences (if you have time feel free to review it and merge).
> 
> I didn't try any zip files yet.

I've tried lots of zip files using the Manage Media page.

**Comment by susanodd** [comment]
> Next is finding out why the files in the zip are not imported.

@vanlummelhuizen as described above, there is a second command that does that.
It is in two parts.

**Comment by vanlummelhuizen** [comment]
OK, uploading a zip and then importing it works, but only after I had to solve some problems.

First, the next two lines delete uploaded videos from a zip before they can be imported. So, the lines should be deleted:

https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/zip_interface.py#L138-L139

Then, there needs to be a slash ("/") after the endpoint `"/dictionary/upload_videos_to_glosses/{datasetid}":

https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/docs/signbank-api.json#L859

Finally, `upload_videos_to_glosses` (see below) needs a csrf_exempt decorator.

https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/api_interface.py#L453-L455

---

Questions:

1. Why two endpoints for one process, that of uploading a zip with videos?
2. Why should the file be downloadable from elsewhere and not via POST request, similar to uploading files on a web page?
3. Why upload a precisely prepared zip when you can upload videos directly with info for dataset, language and/or gloss in the POST request? Since it is an API it is up to the client to handle multiple video uploads.

**Comment by susanodd** [comment]
About the csrf token exception, I have no idea. The code worked fine before the API token and processing was introduced.
It also worked on the signbank-dev server at that time.
I can only assume other changes for using the API token messed it up somehow.
We never put the csrf except on anything before. You only needed to be logged in.
Maybe the swagger is doing something different because it's not actually Signbank.

**Comment by susanodd** [comment]
> About the csrf token exception, I have no idea. The code worked fine before the API token and processing was introduced.
> It also worked on the signbank-dev server at that time.
> I can only assume other changes for using the API token messed it up somehow.
> We never put the csrf except on anything before. You only needed to be logged in.
> Maybe the swagger is doing something different because it's not actually Signbank.

Two endpoints are used to avoid extreme URL access for every single video.
The database gets locked for each video.
The zip stuff just relies on the file system.

Given that, the signbank server starts to "block" when more than 10 videos are processed. It can't keep up with the file system versus all the object creation, making citation images, etc.

The file is downloadable because @susanodd had never written any API at all in 40 years of programming.
Nobody else wanted to work on it at the time.

**Comment by susanodd** [comment]
> About the csrf token exception, I have no idea. The code worked fine before the API token and processing was introduced. It also worked on the signbank-dev server at that time. I can only assume other changes for using the API token messed it up somehow. We never put the csrf except on anything before. You only needed to be logged in. Maybe the swagger is doing something different because it's not actually Signbank.

Swagger isn't logged in, it's only using the API token.

**Comment by susanodd** [comment]
> OK, uploading a zip and then importing it works, but only after I had to solve some problems.
> 
> First, the next two lines delete uploaded videos from a zip before they can be imported. So, the lines should be deleted:
> 
> https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/zip_interface.py#L138-L139
> 
> Then, there needs to be a slash ("/") after the endpoint `"/dictionary/upload_videos_to_glosses/{datasetid}":
> 
> https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/docs/signbank-api.json#L859
> 
> Finally, `upload_videos_to_glosses` (see below) needs a csrf_exempt decorator.
> 
> https://github.com/Signbank/Global-signbank/blob/8c555a05f623ffafb728ecfa65c9cf7cbad2204c/signbank/api_interface.py#L453-L455
> 
> Questions:
> 
>     1. Why two endpoints for one process, that of uploading a zip with videos?
> 
>     2. Why should the file be downloadable from elsewhere and not via POST request, similar to uploading files on a web page?
> 
>     3. Why upload a precisely prepared zip when you can upload videos directly with info for dataset, language and/or gloss in the POST request? Since it is an API it is up to the client to handle multiple video uploads.

The uploaded (unzipped) videos are first put in another location, in the import folder.
That's why the tree is removed there.

**Comment by susanodd** [comment]
@vanlummelhuizen I made the changes for the ZIP code you showed above.
I have the code on signbank-test.

It took a while to figure out what was wrong.
It turns out I had a folder FIRST as a dataset folder under glossvideo.
That was from when i was trying to make a signbank with an empty dataset.

The the dataset folders NEED to be already created in glossvideo.
Otherwise importing the zipped files does not work at all.
Apparently it does not bother to create a dataset folder. (We assume this already exists.)
(The tests also do not check this.)
So once I created the dataset folders, it works now to import a zip file in the Media Manager.

**Comment by susanodd** [comment]
Up to here, the fixes shown above have all been deployed on signbank.

**Comment by susanodd** [comment]
I found this in the logs:
```

```
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/repo/signbank/dictionary/views.py", line 2334, in protected_media
    gloss_pk = int(filename.split('.')[-2].split('-')[-1])
                   ~~~~~~~~~~~~~~~~~~~^^^^
IndexError: list index out of range
Internal Server Error: /dictionary/protected_media/glossvideo/NGT/AA/%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0
%af%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0%af%c0%ae%c0%ae%c0%afetc/passwd
Traceback (most recent call last):
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/repo/signbank/dictionary/views.py", line 2334, in protected_media
    gloss_pk = int(filename.split('.')[-2].split('-')[-1])
                   ~~~~~~~~~~~~~~~~~~~^^^^
IndexError: list index out of range
```

```

**Comment by susanodd** [comment]
```
Internal Server Error: /dictionary/protected_media/handshapeimage/62/
Traceback (most recent call last):
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/repo/signbank/dictionary/views.py", line 2334, in protected_media
    gloss_pk = int(filename.split('.')[-2].split('-')[-1])
                   ~~~~~~~~~~~~~~~~~~~^^^^
IndexError: list index out of range
```

**Comment by susanodd** [comment]
```
[pid: 331|app: 0|req: 71191/368626] 10.208.155.251 () {60 vars in 2445 bytes} [Mon Oct  7 04:51:04 2024] GET /dictionary/protected_media/h
andshapeimage/62/%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%
25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25a
e%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%2
5ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af%25c0%25ae%25c0%25ae%25c0%25af/etc/passwd => generated 90131 bytes in 209 msecs (HTTP
/1.1 500) 5 headers in 185 bytes (1 switches on core 1)
```

Why is there an "/etc/passwd" stuck on the end of the url?
Is this the same hackers that created the fake user accounts?

**Comment by susanodd** [comment]
How to fix this, or is there nothing to do about these malicious  urls?

**Comment by susanodd** [comment]
So if a user tries to enter a bogus URL to a video or image, then try except the only line of code above and do a "Forbidden" ?

**Comment by susanodd** [comment]
```
Forbidden (CSRF token from POST has incorrect length.): /dictionary/handshape/173/
Internal Server Error: /dictionary/protected_media/glossvideo/NGT/TW/%C0%AE%C0%AE/WEB-INF/web.xml
```

**Comment by susanodd** [comment]
Here's more recent, I decided to skip those 200000 lines of hackers.

```
[pid: 848|app: 0|req: 14335/31659] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:07 2024] POST /dictionary/api_update_gloss/
5/47041/ => generated 61 bytes in 79 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 1)
[pid: 848|app: 0|req: 14336/31660] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:07 2024] POST /dictionary/api_update_gloss/
5/47050/ => generated 61 bytes in 88 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 2)
[pid: 844|app: 0|req: 8701/31661] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:07 2024] POST /dictionary/api_update_gloss/5
/47050/ => generated 61 bytes in 95 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 0)
[pid: 844|app: 0|req: 8702/31662] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:07 2024] POST /dictionary/api_update_gloss/5
/47051/ => generated 61 bytes in 86 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 1)
[pid: 844|app: 0|req: 8703/31663] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:08 2024] POST /dictionary/api_update_gloss/5
/47053/ => generated 61 bytes in 82 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
Internal Server Error: /dictionary/api_update_gloss/5/48628/
Traceback (most recent call last):
  File "/var/www/env/lib/python3.12/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: database is locked
```

So after a number of successful API gloss updates, the database is locked.

**Comment by susanodd** [comment]
Here it is in context:

```
[pid: 840|app: 0|req: 5405/31682] 10.208.155.251 () {60 vars in 1255 bytes} [Thu Oct 10 14:39:27 2024] GET /dictionary/package/?extended_f
ields=true&since_timestamp=1728563701 => generated 863033 bytes in 44584 msecs (HTTP/1.1 200) 6 headers in 242 bytes (1 switches on core 3
)
[pid: 840|app: 0|req: 5406/31683] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:12 2024] POST /dictionary/api_update_gloss/5
/47079/ => generated 61 bytes in 92 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 0)
[pid: 848|app: 0|req: 14343/31684] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:12 2024] POST /dictionary/api_update_gloss/
5/47079/ => generated 61 bytes in 76 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 2)
[pid: 844|app: 0|req: 8710/31685] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:12 2024] POST /dictionary/api_update_gloss/5
/49321/ => generated 61 bytes in 69 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
[pid: 840|app: 0|req: 5407/31686] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:12 2024] POST /dictionary/api_update_gloss/5
/47072/ => generated 61 bytes in 116 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 1)
[pid: 839|app: 0|req: 3227/31687] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:13 2024] POST /dictionary/api_update_gloss/5
/44296/ => generated 61 bytes in 108 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 1)
[pid: 848|app: 0|req: 14344/31688] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:13 2024] POST /dictionary/api_update_gloss/
5/47527/ => generated 61 bytes in 89 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 0)
[pid: 848|app: 0|req: 14345/31689] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:13 2024] POST /dictionary/api_update_gloss/
5/46299/ => generated 61 bytes in 76 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
[pid: 848|app: 0|req: 14346/31690] 10.208.155.251 () {62 vars in 1124 bytes} [Thu Oct 10 14:40:13 2024] POST /dictionary/api_update_gloss/
5/46299/ => generated 61 bytes in 74 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 1)
Internal Server Error: /dictionary/api_update_gloss/5/47694/
Traceback (most recent call last):
  File "/var/www/env/lib/python3.12/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/www/env/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: database is locked
```

A package is done, then 8 update gloss calls, then the database is locked.

**Comment by susanodd** [comment]
Question: Can we force the API calls to fail if the database is locked? Otherwise the stuff afterwards just keeps causing errors. Django ought to refuse new API requests.

**Comment by susanodd** [comment]
@vanlummelhuizen what is the best kind of response to "revoke" all of those "hacker" type url requests? https://github.com/Signbank/Global-signbank/issues/1331#issuecomment-2407256811
They ought to be revoked before the server tries to process them. Are there Django settings? It's a bunch of bogus stuff stuck on at the end of legitimate urls.

**Comment by susanodd** [comment]
I removed the bogus user accounts that showed up in the database last week.
It was pieces of code used as user names.

**Comment by susanodd** [comment]
```
[sqlite3.connect](https://docs.python.org/3.8/library/sqlite3.html#sqlite3.connect) has an optional isolation_level parameter, which controls the database locking behavior during transactions.

In DEFERRED (the default) mode, a transaction acquires a read lock on the database when it begins, and a write lock only when a write operation is requested.
In IMMEDIATE mode, a transaction acquires a write lock immediately, even if it's only performing read operations. Other transactions attempting to write to the database will be blocked, but read-only transactions can still proceed.
In EXCLUSIVE mode, a transaction acquires an exclusive lock on the database, blocking all other transactions, reading or writing, until the current transaction is committed or rolled back.
```

**Comment by susanodd** [comment]
> From [Django's documentation](https://docs.djangoproject.com/en/dev/ref/databases/#database-is-locked-errors):
> 
> SQLite is meant to be a lightweight database, and thus can't support a high level of concurrency. OperationalError: database is locked errors indicate that your application is experiencing more concurrency than sqlite can handle in default configuration. This error means that one thread or process has an exclusive lock on the database connection and another thread timed out waiting for the lock the be released.

**Comment by vanlummelhuizen** [comment]
About moving away from SQLite: #673 

About request from hackers: Django seems to handle those correctly, responding with error 500. If you want to know more about security (hardening) search for 'django security hardening'. One thing I came across that might be of use: https://djcheckup.com/.

About bogus user registrations: we propably needs to review registrations once in a while. But I think it should be an Amsterdam task. Perhaps we could make a cronjob that make a daily email with al new registration and send that to Ulrika for instance.

**Comment by susanodd** [comment]
Here is where the database is with the bogus users still in it (tensusers):  `signbank_101024.db`

**Comment by susanodd** [comment]
@vanlummelhuizen do I need to add a try-except on all the save methods? And throw an error to the API if the database is locked?

I don't know if methods that fail clean up after themselves.
